board/main.c

```diff
@@ -192,14 +192,69 @@ int get_rtc_pkt(void *dat) {
   return sizeof(t);
 }
 
+typedef struct {
+  uint8_t ptr;
+  uint8_t tail_size;
+  uint8_t data[72];
+  uint8_t counter;
+} usb_asm_buffer;
+
+usb_asm_buffer ep1_buffer = {.ptr = 0, .tail_size = 0, .counter = 0};
+uint32_t total_rx_size = 0;
+
 int usb_cb_ep1_in(void *usbdata, int len, bool hardwired) {
   UNUSED(hardwired);
-  CAN_FIFOMailBox_TypeDef *reply = (CAN_FIFOMailBox_TypeDef *)usbdata;
-  int ilen = 0;
-  while (ilen < MIN(len/0x10, 4) && can_pop(&can_rx_q, &reply[ilen])) {
-    ilen++;
+  uint8_t pos = 1;
+  uint8_t *usbdata8 = (uint8_t *)usbdata;
+  // Send tail of previous message if it is in buffer
+  if (ep1_buffer.ptr > 0) {
+    if (ep1_buffer.ptr <= 63U) {
+      (void)memcpy(&usbdata8[pos], ep1_buffer.data, ep1_buffer.ptr);
+      pos += ep1_buffer.ptr;
+      ep1_buffer.ptr = 0;
+    } else {
+      (void)memcpy(&usbdata8[pos], ep1_buffer.data, 63U);
+      ep1_buffer.ptr = ep1_buffer.ptr - 63U;
+      (void)memcpy(ep1_buffer.data, &ep1_buffer.data[63U], ep1_buffer.ptr);
+      pos += 63U;
+    }
+  }
+
+  if (total_rx_size > MAX_EP1_CHUNK_PER_BULK_TRANSFER) {
+    if (pos > 1) { usbdata8[0] = ep1_buffer.counter; }
+    total_rx_size = 0;
+    ep1_buffer.counter = 0;
+  } else {
+    CANPacket_t can_packet;
+    while ((pos < len) && can_pop(&can_rx_q, &can_packet)) {
+      uint8_t pckt_len = CANPACKET_HEAD_SIZE + dlc_to_len[can_packet.data_len_code];
+      if ((pos + pckt_len) <= len) {
+        (void)memcpy(&usbdata8[pos], &can_packet, pckt_len);
+        pos += pckt_len;
+      } else {
+        (void)memcpy(&usbdata8[pos], &can_packet, len - pos);
+        ep1_buffer.ptr = pckt_len - (len - pos);
+        //(void)memcpy(ep1_buffer.data, ((uint8_t*)&can_packet + (len - pos)), ep1_buffer.ptr);
+        // cppcheck-suppress objectIndex
+        (void)memcpy(ep1_buffer.data, &((uint8_t*)&can_packet)[(len - pos)], ep1_buffer.ptr);
+        pos = len;
+      }
+    }
+    if (pos > 1) {
+      usbdata8[0] = ep1_buffer.counter;
+      ep1_buffer.counter++;
+    }
+    else {
+      pos = 0;
+      ep1_buffer.counter = 0;
+    }
+    if (pos != len) {
+      ep1_buffer.counter = 0;
+      total_rx_size = 0;
+    }
+    total_rx_size += pos;
```

Comment:
let's add some comments here, kinda hard to follow the logic like this

Reply:
can't easily tell what this does or what it's supposed to do. Just stitch together a single message larger than the 64 byte frame?

Reply:
also: do we want to compile this out and add some unit tests to this to make sure it's good in all edge cases?

Reply:
Big fan of putting all the ring buffer code in a tested library!